internal/test/crds.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gobuffalo/flect"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+var (
+	bootstrapGroupVersion = schema.GroupVersion{Group: "bootstrap.cluster.x-k8s.io", Version: "v1alpha4"}
+
+	// BootstrapCRD is a generic boostrap CRD.
+	BootstrapCRD = generateCRD(bootstrapGroupVersion.WithKind("Bootstrap"))
+
+	// BootstrapTemplateCRD is a generic boostrap template CRD.
+	BootstrapTemplateCRD = generateCRD(bootstrapGroupVersion.WithKind("BootstrapTemplate"))
+
+	controlPlaneGroupVersion = schema.GroupVersion{Group: "controlplane.cluster.x-k8s.io", Version: "v1alpha4"}
+
+	// ControlPlaneCRD is a generic control plane CRD.
+	ControlPlaneCRD = generateCRD(controlPlaneGroupVersion.WithKind("ControlPlane"))
+
+	// ControlPlaneTemplateCRD is a generic control plane template CRD.
+	ControlPlaneTemplateCRD = generateCRD(controlPlaneGroupVersion.WithKind("ControlPlaneTemplate"))
+
+	infrastructureGroupVersion = schema.GroupVersion{Group: "infrastructure.cluster.x-k8s.io", Version: "v1alpha4"}
+
+	// InfrastructureMachineCRD is a generic infrastructure machine CRD.
+	InfrastructureMachineCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureMachine"))
+
+	// InfrastructureMachineTemplateCRD is a generic infrastructure machine template CRD.
+	InfrastructureMachineTemplateCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureMachineTemplate"))
+
+	// InfrastructureClusterCRD is a generic infrastructure machine CRD.
+	InfrastructureClusterCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureCluster"))
+
+	// InfrastructureClusterTemplateCRD is a generic infrastructure machine template CRD.
+	InfrastructureClusterTemplateCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureClusterTemplate"))
+
+	// InfrastructureRemediationCRD is a generic infrastructure remediation CRD.
+	InfrastructureRemediationCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureRemediation"))
```

Comment:
I moved the remediation resources to v1alpha4 to avoid creating another infrastructureGroupVersion with v1alpha3. If we prefer to keep them in v1alpha3, I'll change it of course.

Reply:
+1

Reply:
Also this is not really required to be part of an infrastructure provider, should it be `GenericExternalRemediation`